WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsConnectedViewModel.swift

```diff
@@ -58,11 +60,19 @@ final class CardReaderSettingsConnectedViewModel: CardReaderSettingsPresentedVie
     ///
     func disconnectReader() {
         ServiceLocator.analytics.track(.cardReaderDisconnectTapped)
-        let action = CardPresentPaymentAction.disconnect() { result in
+
+        let readerSerialNumberToDisconnect = connectedReaderSerialNumber
```

Comment:
I am not sure I understand this block. If I understand correctly, disconnect is always called, but forgetCardReader is only called, after disconnect is completed, if the serial number is not nil?

Is there a reason why we can't guard against the serial number of the connected reader and call disconnect only when the serial number is not nil? If not, could the call to forgetCardReader be encapsulated in a separate method? This new method would be the one that guards against a nil serial number.

The reason I ask this is because this line seems a bit out of place in the context of this codebase: 
`[readerSerialNumberToDisconnect, weak self]`

Reply:
Hmmm. My thoughts were to forget a card reader only after successfully disconnecting from it, to avoid the possibility of introducing an odd state - where we are connected to a reader, but it is not longer known/remembered.

> Is there a reason why we can't guard against the serial number of the connected reader and call disconnect only when the serial number is not nil?

Perhaps, but the view model would then need to know what's important in that CardReader that was passed to it. I think that detail should remain known only to the lower levels.

In fact, as I type that, I'm going to rename the "Serial number" variables to use simply "ID" like I do elsewhere. The fact that we are using the serial number as an ID is something I should be getting from the CardReader object. I think I'll add/use a helper on that object to get the ID.

> If not, could the call to forgetCardReader be encapsulated in a separate method? This new method would be the one that guards against a nil serial number.

Can do

Reply:
OK. I opted for simply forgetting the reader whenever disconnect is called. In the very unlikely event disconnect fails, the next time the merchant connects to the reader we'll save the readerID anyways. The unusual `[readerSerialNumberToDisconnect, weak self]` is now no longer needed.

I also cleaned up the use of serial number to be simply "ID"

Changes made in https://github.com/woocommerce/woocommerce-ios/pull/4526/commits/66195bc396839ad43d340cf9dfa2b0e6546f9326